src/main/groovy/de/dkfz/roddy/execution/io/LocalExecutionHelper.groovy

```diff
@@ -117,23 +146,51 @@ class LocalExecutionHelper {
      */
     static ExecutionResult executeCommandWithExtendedResult(
             String command,
+            Boolean waitFor,
+            Duration timeout = Duration.ZERO,
             OutputStream outputStream = null,
             ExecutorService executorService = executorService) {
         List<String> bashCommand = ["bash", "-c", command]
-        logger.postRareInfo("Executing the command ${command} locally.")
+        logger.postRareInfo("Executing the command ${bashCommand} locally.")
         ProcessBuilder processBuilder = new ProcessBuilder(bashCommand)
         Process process = processBuilder.start()
-        Future<List<String>> stdout
+        Future<List<String>> stdoutF
         if (outputStream == null)
-            stdout = asyncReadStringStream(process.inputStream, executorService)
+            stdoutF = asyncReadStringStream(process.inputStream, executorService)
         else
-            stdout = asyncReadStringStream(process.inputStream, executorService, outputStream)
-        Future<List<String>> stderr = asyncReadStringStream(process.errorStream, executorService)
-        Future<Integer> exitCode = CompletableFuture.supplyAsync({
-            process.waitFor()
+            stdoutF = asyncReadStringStream(process.inputStream, executorService, outputStream)
+        Future<List<String>> stderrF =
+                asyncReadStringStream(process.errorStream, executorService)
+        Future<Integer> exitCodeF = CompletableFuture.supplyAsync({
+            Integer result
+            if (timeout != Duration.ZERO) {
+                // To ensure the command is actually terminated and for better error reporting
+                // a TimeoutException is raised here, rather than creating a timeout at the level
+                // of the future.
+                Boolean finished = process.waitFor(timeout.toNanos(), TimeUnit.NANOSECONDS)
+                if (finished) {
+                    result = process.exitValue()
+                } else {
+                    // Supplier.get() sucks. It doesn't declare throwing exceptions. Without the
+                    // wrapping one will get an UndeclaredThrowableException. Let's do this
+                    // explicitly into an unchecked Exception, instead.
+                    throw new RuntimeException(new TimeoutException("Command execution timed out: ${bashCommand}"))
+                }
+            } else {
+                result = process.waitFor()
+            }
+            result
         } as Supplier<Integer>, executorService)
-        return new AsyncExecutionResult(bashCommand, getProcessID(process),
-                exitCode, stdout, stderr).asExecutionResult()
+
+        AsyncExecutionResult result = new AsyncExecutionResult(
+                bashCommand,
+                getProcessID(process),
+                exitCodeF, stdoutF, stderrF)
+        if (waitFor) {
+            return result.asExecutionResult()
+        } else {
```

Comment:
else is unnecessary here, as there is a return.

Reply:
I renamed to `asSynchronousExecutionResult`. The differences in the if-else branches is that that if `waitFor` is set, the `resultMethod` wait's for the execution's actual result. Otherwise the asynchronous result object is returned. The else branch therefore makes sense.

In any case, this code is not nice and mainly for backwards compatibility, because one of the methods for `executeCommandWithExtendedResult` is used at different places in Roddy. As soon as RoddyToolLib and BE are not used in Roddy anymore. In WESkit similar code differentiates between asynchronous process handles (with some information about the process) and results that exist only after the command finished.